Add Clifford circuit optimization to BasicSimulator

debasmita2102 opened the pull request:
- Implement _is_clifford_circuit() to detect Clifford circuits
- Implement _run_clifford_circuit() using StabilizerState for efficient simulation
- Fixes #13610

debasmita2102 committed:
Add Clifford circuit optimization to BasicSimulator

- Implement _is_clifford_circuit() to detect Clifford circuits
- Implement _run_clifford_circuit() using StabilizerState for efficient simulation
- Fixes #13610

ShellyGarion commented:
Any idea why CI tests python 3.13 look stuck? @kt474?

alexanderivrii commented on qiskit/providers/basic_provider/basic_simulator.py:
```diff
@@ -500,6 +505,116 @@ def _run_job(self, job_id: str, run_input) -> Result:
+
+        # Create circuit without measurements
+        circ_no_meas = circuit.copy()
+        circ_no_meas.remove_final_measurements()
```
Looking at the docs for `QuantumCircuit.remove_final_measurements` (see https://quantum.cloud.ibm.com/docs/en/api/qiskit/qiskit.circuit.QuantumCircuit), the method accepts the argument `inplace`, so you can simplify the above two lines to:

```suggestion
        circ_no_meas = circuit.remove_final_measurements(inplace=False)
```

More importantly, the docs also state that the method has rather complex behavior (something about removal of idle classical bits) and suggest to use the transpiler pass `RemoveFinalMeasurements` instead if possible (providing an explicit example on how to do so). Which of the two do you think we need here?

debasmita2102 replied:
Thank! I'll use `remove_final_measurements(inplace=False)` for this case because:  It's cleaner as a one-liner. We are just temporarily removing measurements for the Clifford check, not doing full circuit optimization, so the transpiler pass may be overkill?? 
I've updated the code to use `circuit.remove_final_measurements(inplace=False)` as suggested.

alexanderivrii commented on qiskit/providers/basic_provider/basic_simulator.py:
```diff
@@ -535,6 +650,17 @@ def _run_circuit(self, circuit) -> dict:
+        if self._is_clifford_circuit(circuit):
+            # Only use Clifford optimization if no custom initial state
+            if self._initial_statevector is None:
+                return self._run_clifford_circuit(circuit)
```
Several comments about this code. Converting a quantum circuit to a Clifford is not free (which is what the function `_is_clifford_circuit` does). 

Thus, first of all, it makes sense to change the order of the two if statements, so that we don't need to do the more expensive computation when `self._initial_statevector` is not `None`. Actually, why do we need this restriction?

Second, the code seems to report the time taken by the simulation, so the time for the above check must be also measured and taken into account. 

Third, calling `stab_state = StabilizerState(circ_no_meas)` essentially does exactly the same work as calling `Clifford(circ_no_meas)`, so you are essentially doing the same work twice. You could potentially change the output of `_is_clifford_circuit` to be a Clifford or None (depending on whether the circuit is Clifford or not), and when not None, get the stabilizer state from Clifford without the need for recomputation.

debasmita2102 replied:
I've implemented all three suggestions:

**1. Reordered if statements:**
Now checking `self._initial_statevector is None` first before doing the more expensive Clifford detection. This avoids unnecessary work when a custom initial state is set.

**2. Timing consideration:**
The Clifford detection time is now properly accounted for since it happens within `_run_circuit` before returning. The timing measurement captures the full execution path.

**3. Eliminated double Clifford construction:**

- `_is_clifford_circuit(circuit)` now returns the `Clifford` object itself (or `None` if not Clifford)
- `_run_clifford_circuit(circuit, clifford_obj)` accepts and reuses that pre-computed object
- This eliminates the duplicate `Clifford(circ_no_meas)` construction, cutting overhead significantly

**Regarding the `_initial_statevector` restriction:**
I kept this check because `StabilizerState` can only be initialized from:
- A `Clifford` circuit/object
- A computational basis state string (e.g., "000")
- Another `StabilizerState`

It doesn't accept arbitrary complex statevectors like `[0.5, 0.5, 0.5, 0.5]`, so we can't use the Clifford optimization when the user provides a custom initial state. Should I document this limitation more clearly in the code?

alexanderivrii replied:
Thanks for the explanation!

alexanderivrii reviewed (commented):
Thank you for working on this. Intuitively I completely agree that using the stabilizer state formalism (when possible) should be significantly more efficient than multiplying unitary matrices. Do you have some experimental data showing the improvement in performance?

I also left some preliminary comments inline, in particular there are some immediate performance improvements that you can do.

debasmita2102 commented:
Thank you for the detailed review, @alexanderivrii. I have addressed all the feedback and ran comprehensive benchmarks.

### Performance Benchmarks

After implementing your code review suggestions, here are the results:

**Clifford circuits (100 shots):**
Qubit Count  |  Time (seconds)
-------------+----------------
10 qubits    |  0.312s        
15 qubits    |  0.918s        
20 qubits    |  2.090s        
22 qubits    |  2.706s        


**Comparison with non-Clifford path:**
Testing the same circuits with a T gate added (forcing statevector path) shows nearly identical performance (within ~1-2%). Both paths hit BasicSimulator's `MAX_QUBITS_MEMORY = 24` limit at the same point.

### Analysis

The theoretical O(n²) vs O(2ⁿ) scaling advantage doesn't materialize significantly because:
- Both paths hit the same qubit limit (~24)
- Clifford detection overhead roughly equals simulation savings
- BasicSimulator's statevector operations are already efficient for small circuits

The optimization provides clean structure and automatic Clifford routing without hurting performance, but speedups are **modest to negligible** for typical BasicSimulator use cases.

### Code Review Items Addressed

✅ All inline suggestions implemented:
1. Simplified `remove_final_measurements(inplace=False)` 
2. Reordered checks (`_initial_statevector` first)
3. Eliminated double Clifford construction by returning and reusing the object

Let me know what else is left.

debasmita2102 committed:
Address code review feedback from alexanderivrii

- Use remove_final_measurements(inplace=False) for cleaner code
- Reorder checks: test initial_statevector first (cheaper operation)
- Refactor to avoid double Clifford construction by returning Clifford
  object from _is_clifford_circuit and reusing in _run_clifford_circuit

debasmita2102 committed:
Update docstring for _run_clifford_circuit

Add missing clifford_obj parameter documentation to docstring.

debasmita2102 committed:
Merge branch 'main' into fix/issue-13610-stabilizer-backend

debasmita2102 committed:
Merge branch 'main' into fix/issue-13610-stabilizer-backend

ShellyGarion reviewed (dismissed):
LGTM, thanks @debasmita2102 for handling this!
I think that the Clifford simulator should work for more than 24 qubits (there should be no problem to generate a Clifford on 100 qubits), but since this is a hard coded constant we don't have to handle it now.

```
    # Formerly calculated as `int(log2(local_hardware_info()["memory"]*(1024**3)/16))`.
    # After the removal of `local_hardware_info()`, it's hardcoded to 24 qubits,
    # which matches the ~268 MB of required memory.
    MAX_QUBITS_MEMORY = 24
```

alexanderivrii reviewed (dismissed):
Thank you for addressing my comments!

ShellyGarion committed:
Merge branch 'main' into fix/issue-13610-stabilizer-backend

ShellyGarion commented:
This PR was removed from merge queue since `Python Unit Tests (3.13, windows-latest) / windows-latest-tests-Python-3.13 ` are stalling. This happened several times during the CI checks of this PR. I have no idea why this happens.

jakelishman commented on qiskit/providers/basic_provider/basic_simulator.py:
```diff
@@ -500,6 +505,98 @@ def _run_job(self, job_id: str, run_input) -> Result:
+            # Sample each shot individually
+            for _ in range(self._shots):
+                # Create fresh StabilizerState for each shot
+                stab_state = StabilizerState(clifford_obj)
```
This recalculates the entire `StabilizerState` on every shot.  Calculating the stabilizer state is expensive, and doesn't change between shots.  At a minimum, you can save the state outside of the loop, and copy it within the loop if you need to mutate it, but also, I think the only mutation to it is to set the seed.  Can we reset the seed of the same instance in each iteration?

alexanderivrii replied:
I thought that constructing a stabilizer state from a Clifford should be fast, but now I see that it does copy the symplectic tableau on every call, and furthermore `StabilizerState.__init__` has `validate=True` by default, which seems to do quite a bit of work (we should definitely call with `validate=False`).

debasmita2102 replied:
Updates: Creating `StabilizerState` once, Moved outside the loop with `validate=False`. 

Used Reset `_rng.seed()` each iteration instead of copying.

jakelishman commented on qiskit/providers/basic_provider/basic_simulator.py:
```diff
@@ -500,6 +505,98 @@ def _run_job(self, job_id: str, run_input) -> Result:
 
         return Result.from_dict(result)
 
+    def _is_clifford_circuit(self, circuit: QuantumCircuit):
+        """Check if circuit is Clifford and return Clifford object or None."""
+        try:
+            circ_no_meas = circuit.remove_final_measurements(inplace=False)
+            return Clifford(circ_no_meas)
+        except QiskitError:
+            return None
```
This is really a rather expensive check, and it runs on every circuit - `remove_final_measurements` is not cheap, nor is constructing a Clifford object.  We probably need to revisit this.

debasmita2102 replied:
I have made it opt-in via the `use_clifford_optimization` flag (default: `False`), so it only runs when users explicitly enable Clifford optimization.

Though there is still room for improvement even when the flag is enabled because`remove_final_measurements()` creates a copy and  `Clifford()` construction fails slowly for non-Clifford circuits

ShellyGarion replied:
what happens if the user provides `use_clifford_optimization = True` but the circuit is not Clifford? 
does it use the usual `BasicSimulator` instead or does it raise an error message? 
(perhaps it's worth to add a test for this case)

jakelishman commented on qiskit/providers/basic_provider/basic_simulator.py:
```diff
@@ -535,6 +632,18 @@ def _run_circuit(self, circuit) -> dict:
+        self._number_of_qubits = circuit.num_qubits
+        self._number_of_cmembits = circuit.num_clbits
+
+        # Check if circuit is Clifford and use optimized simulation
+        # Check initial_statevector first (cheaper)
+        if self._initial_statevector is None:
+            clifford_obj = self._is_clifford_circuit(circuit)
+            if clifford_obj is not None:
+                return self._run_clifford_circuit(circuit, clifford_obj)
```
Most likely, this is being very very expensive, especially when called in a loop.  The test for "is this a Clifford circuit?" fails very slowly for circuits that _aren't_ Cliffords.

ShellyGarion replied:
@jakelishman - is there some parameter that one could give the `BasicSimulator` to indicate that this is a Clifford circuit? and only in this case one would do the checks for a Clifford circuit?

In addition, in the case of a Clifford circuit, is it possible to change the default value (=24) of `MAX_QUBITS_MEMORY` ?

debasmita2102 replied:
@jakelishman I. have made the clifford optimization opt-in. 
-Added `use_clifford_optimization` option (default: `False`)
-The expensive `_is_clifford_circuit()` check only runs when users explicitly enable it
Example usage:
```
backend = BasicSimulator(use_clifford_optimization=True)
result = backend.run(clifford_circuit, shots=1000)
```

This way, the optimization is available for users who knows they have Clifford circuits, but doesn't impact general use cases.

debasmita2102 replied:
@ShellyGarion Currently `MAX_QUBITS_MEMORY=24` applies to both statevector and Clifford paths. Would you like me to add a separate qubit limit for Clifford circuits or to make this configurable via an option? Or to keep it as-it-is for now?

ShellyGarion replied:
Could you perhaps calculate the bound for the maximal number of qubits for Clifford circuits based on this [comment](https://github.com/Qiskit/qiskit/pull/15159#pullrequestreview-3421331081)?

How would you suggest to update the hard-coded constant of `MAX_QUBITS_MEMORY=24` for Clifford circuits?

ShellyGarion replied:
According to a rule of thumb, a Clifford tableau on N qubits contains 2Nx2N elements, while a state vector on M qubits has 2^M complex elements. 
So if we have a memory of 2^24 element, we can use it to store a Clifford tableau on more than 2000 qubits (=sqrt(2^24)/2). 
Does it make sense?

debasmita2102 replied:
#### Maximum Qubits Calculation

**Memory budget:** 2^24 elements

**For Clifford tableau:**
- Size: 2N × 2N elements
- 2^24 = 2N × 2N = 4N²
- N² = 2^24 / 4 = 2^22
- N = sqrt(2^22) = 2^11 = **2048 qubits**

**For statevector:**
- Size: 2^M complex elements
- 2^24 = 2^M
- M = **24 qubits**

So Clifford circuits can theoretically handle **2048 qubits** vs only **24 qubits** for statevector!


#### Implementation Approach

I can add a separate constant for Clifford circuits:

**Proposed code:**
```
class BasicSimulator(BackendV2):
MAX_QUBITS_MEMORY = 24 # Statevector limit (~268 MB)
MAX_QUBITS_CLIFFORD = 2048 # Clifford tableau (2N×2N scaling)

def _run(self, run_input, **options):
    # ... existing setup code ...
    
    # Check Clifford optimization path first
    if self._use_clifford_optimization:
        clifford_obj = self._is_clifford_circuit(circuit)
        if clifford_obj is not None:
            # Validate Clifford qubit limit
            if circuit.num_qubits > self.MAX_QUBITS_CLIFFORD:
                raise BasicProviderError(
                    f"Number of qubits ({circuit.num_qubits}) exceeds "
                    f"Clifford limit ({self.MAX_QUBITS_CLIFFORD})"
                )
            return self._run_clifford_circuit(circuit, clifford_obj)
    
    # Standard statevector path
    if circuit.num_qubits > self.MAX_QUBITS_MEMORY:
        raise BasicProviderError(
            f"Number of qubits ({circuit.num_qubits}) exceeds "
            f"statevector limit ({self.MAX_QUBITS_MEMORY})"
        )
    # ... continue with statevector simulation ...
```

Should I implement this approach? This would finally let us see if the optimization provides speedup in the regime where it's supposed to...

ShellyGarion replied:
that might be a good approach, and perhaps worth to try, as currently it seems that the stabilizer simulator is worse than the statevector simulator for <= 24 qubits. 
I think that we should be more careful with the size limits since the simulator keeps copies of the intrnal clifford during the measurement calculation, so maybe we use 8*N^3 memory during the process, but it's still better than 2^N.

jakelishman reviewed (commented):
This PR is causing the runtimes of several tests to explode - the logs indicate that these tests are the slowest:
```
test.python.primitives.test_backend_estimator_v2.TestBackendEstimatorV2.test_run_single_circuit_observable_4                  264.740
test.python.primitives.test_backend_estimator_v2.TestBackendEstimatorV2.test_run_single_circuit_observable_3                  249.959
test.python.primitives.test_backend_estimator_v2.TestBackendEstimatorV2.test_run_2qubits_4                                    198.009
test.python.primitives.test_backend_estimator_v2.TestBackendEstimatorV2.test_run_2qubits_3                                    186.913
test.python.primitives.test_backend_estimator_v2.TestBackendEstimatorV2.test_run_1qubit_4                                     137.465
test.python.primitives.test_backend_estimator_v2.TestBackendEstimatorV2.test_run_1qubit_3                                     135.929
```

all of which are using `BasicSimulator` as the backend.  None of those appear in the slowest tests of other PRs at all, so it's probably this PR causing the slowdown.

I had a quick skim through and highlighted some places that might be the cause - we should look through these and ensure we're not significantly regressing on performance on them before merge, since this PR is supposed to speed things up.

debasmita2102 committed:
Make Clifford optimization opt-in and fix StabilizerState performance

Add use_clifford_optimization flag (default False) and optimize state creation

ShellyGarion reviewed (commented):
Hi @debasmita2102 - thanks for applying all these cahnges, I have a few further questions:

1. Could you please compare the tests in this [comment](https://github.com/Qiskit/qiskit/pull/15159#pullrequestreview-3428314351) for the main branch and your PR to see that there are no performence degregation?

2. Could you please repeat the benchmarks comaring the `BasicSimulator` for Clifford and non-Clifford circuits (for ~20 qubits) and compare their performence now?

debasmita2102 committed:
Improve Clifford optimization implementation

- Sample all shots at once with sample_memory()
- Create StabilizerState once per run
- Fix RNG seeding

debasmita2102 commented:
What Changed in a269688
1. **Eliminated per-shot overhead**: Now use `sample_memory(self._shots)` to sample all 1000 shots at once, instead of looping 1000 times
2. **Single StabilizerState creation**: Create once with `validate=False` outside the loop
3. **Fixed RNG seeding**: Correctly call `stab_state.seed()` after creation


**Before (slow):**
```
for shot in range(1000):
stab_state.seed(shot) # 1000 expensive seed calls
sample = stab_state.sample_memory(1)

```


**After (optimized):**
```
stab_state = StabilizerState(clifford_obj, validate=False)
stab_state.seed(self._seed_simulator)
samples = stab_state.sample_memory(1000) # One call!
```


These changes build on the opt-in design.

debasmita2102 commented:
@ShellyGarion  Your question : **What happens if user provides `use_clifford_optimization=True` but the circuit is not Clifford:**

The code falls back to regular statevector simulation
Eg: 
Non-Clifford circuit (T gate makes it non-Clifford)
```
qc = QuantumCircuit(3, 3)
qc.h(0)
qc.t(1) # T gate is non-Clifford
qc.cx(0, 1)
qc.measure(, )
```

Enable optimization anyway:
```
backend = BasicSimulator(use_clifford_optimization=True)
# This should NOT error
print("Running non-Clifford circuit with use_clifford_optimization=True...")
result = backend.run(non_clifford_qc, shots=100).result()
print("SUCCESS! No error raised.")  
print(f"Counts: {result.get_counts()}")
```


**Output:**
Running non-Clifford circuit with use_clifford_optimization=True
SUCCESS! No error raised.
Counts: {'000': 48, '011': 52}

debasmita2102 commented:
> Hi @debasmita2102 - thanks for applying all these cahnges, I have a few further questions:
> 
> 1. Could you please compare the tests in this [comment](https://github.com/Qiskit/qiskit/pull/15159#pullrequestreview-3428314351) for the main branch and your PR to see that there are no performence degregation?



@ShellyGarion Test suite performance comparison:

**Main branch:**
- test_run_single_circuit_observable_1: 0.47s
- test_run_2qubits_4: 0.17s
- test_run_1qubit_3: 0.07s
- **Total: 0.98s**

**PR branch (opt-in OFF):**
- test_run_single_circuit_observable_1: 0.45s
- test_run_2qubits_4: 0.17s
- test_run_1qubit_3: 0.07s
- **Total: 0.98s**

NOTE: opt-in flag is OFF by default.

> 2. Could you please repeat the benchmarks comaring the `BasicSimulator` for Clifford and non-Clifford circuits (for ~20 qubits) and compare their performence now?

Benchmark results for Clifford vs non-Clifford circuits (1000 shots):

| Qubits | Non-Clifford | Clifford (opt OFF) | Clifford (opt ON) | Speedup |
|--------|--------------|-------------------|-------------------|---------|
| 10       | 0.003s          | 0.002s                        | 0.717s                      | 0.00x  |
| 15       | 0.014s           | 0.011s                        | 1.413s                      | 0.01x  |
| 20       | 0.454s           | 0.336s                     | 2.328s                      | 0.14x  |

**Key findings:**
- Default behavior unchanged (no regression)
- Optimization slower for < 24 qubits due to Clifford detection overhead
-  Can't test 25+ qubits (MAX_QUBITS_MEMORY=24 limit)

**Questions:**
1. Should we increase MAX_QUBITS_MEMORY for Clifford circuits?
2. Is there a cheaper Clifford detection method?

debasmita2102 commented:
@jakelishman  @ShellyGarion  The Clifford optimization is adding significant overhead rather than improving performance. The issue is that `_is_clifford_circuit()` creates a full Clifford object just to detect whether the circuit is Clifford, circuits that normally complete in 0.002-0.336s instead take 0.7-2.3s. This detection cost far exceeds any potential speedup from using StabilizerState instead of statevector simulation for circuits under 24 qubits. I guess this optimization would theoretically benefit larger circuits (30+ qubits) where statevector becomes exponentially expensive, but BasicSimulator has a hard MAX_QUBITS_MEMORY limit of 24 qubits. I attempted to test 25-qubit circuits and hit the limit immediately. So we're stuck. The opt-in design ensures no regression since the expensive code path only runs when explicitly enabled, but the feature has limited practical value unless the qubit limit is increased or a cheaper Clifford detection method is implemented.

alexanderivrii commented:
Hi @debasmita2102, you could have a simpler check whether a circuit is Clifford just by examining the gate names. This approach would miss certain circuits (for instance, RZ(pi/2) gate is Clifford, but does not have a "Clifford" gate name), however maybe this would be good enough in practice. However, you would still need to construct a `StabilizerState`, correct? And if I remember correctly this still calls the same code as the code in `_is_clifford_circuit`. Can you give an explicit example of a Clifford circuit where you get significantly worse runtimes with OPT=ON?

debasmita2102 commented:
@alexanderivrii 
> Can you give an explicit example of a Clifford circuit where you get significantly worse runtimes with OPT=ON?
Code to run:

```
import time
from qiskit import QuantumCircuit
from qiskit.providers.basic_provider import BasicSimulator

# Create a Clifford circuit (20 qubits)
print("Creating 20-qubit Clifford circuit...")
qc = QuantumCircuit(20, 20)
for i in range(19):
    qc.h(i)
    qc.cx(i, i+1)
qc.measure_all()

print(f"Circuit has {qc.num_qubits} qubits and {qc.depth()} depth\n")

# Test with OPT=OFF (default)
print("=" * 60)
print("Test 1: OPT=OFF (default statevector simulation)")
print("=" * 60)
backend_off = BasicSimulator()
start = time.time()
result_off = backend_off.run(qc, shots=1000).result()
time_off = time.time() - start
print(f"Time: {time_off:.3f}s")
print(f"Sample counts: {dict(list(result_off.get_counts().items())[:3])}")

# Test with OPT=ON
print("\n" + "=" * 60)
print("Test 2: OPT=ON (Clifford optimization)")
print("=" * 60)
backend_on = BasicSimulator(use_clifford_optimization=True)
start = time.time()
result_on = backend_on.run(qc, shots=1000).result()
time_on = time.time() - start
print(f"Time: {time_on:.3f}s")
print(f"Sample counts: {dict(list(result_on.get_counts().items())[:3])}")

# Comparison
print("\n" + "=" * 60)
print("COMPARISON")
print("=" * 60)
print(f"OPT=OFF: {time_off:.3f}s")
print(f"OPT=ON:  {time_on:.3f}s")
print(f"Slowdown: {time_on/time_off:.1f}x slower with optimization")

```
Results of slow down ---
```
**Results:**
OPT=OFF: 0.379s
OPT=ON: 0.683s
Slowdown: 1.8x slower with optimization
```

alexanderivrii commented:
Thanks @debasmita2102. Running cProfile profiler on your benchmark, shows that most of the runtime is actually _not_ spent on constructing `Clifford`/`StabilizerState` from the circuit, but on the function `sample_memory` in `stabilizer_state`. 

<img width="1225" height="434" alt="Screenshot 2025-11-17 at 16 47 55" src="https://github.com/user-attachments/assets/bfb46e70-3b9c-4bda-98d1-dd38687f1455" />

I believe that might be some potential to optimize the code even without porting it to Rust. One immediate thing: both `sample_memory` and `measure` in `stabilizerstate.py` seem to copy the given stabilizer state object, and I believe one of these copies can be removed. I am also surprised that `_measure_and_update` takes such a long time.

alexanderivrii commented:
Hi, to be clear: I am not necessarily suggesting to improve the performance of StabilizerState simulation as a part of this PR, I think this deserves a PR on its own. In addition, @ShellyGarion has been working on porting some of the Clifford code into rust, which might be the best longer-term solution.

ShellyGarion commented:
> Hi, to be clear: I am not necessarily suggesting to improve the performance of StabilizerState simulation as a part of this PR, I think this deserves a PR on its own. In addition, @ShellyGarion has been working on porting some of the Clifford code into rust, which might be the best longer-term solution.

I agree, performance issues in `StabilizerState` code should be handled sperately, perhaps when porting the code to rust.

ShellyGarion commented:
Note that the benchmark in this [comment](https://github.com/Qiskit/qiskit/pull/15159#issuecomment-3540772186), is a state with 2^{N-1} possible outputs. What happens if we take a GHZ circuit which has only 2 possible outputs? or big random clifford circuits (https://quantum.cloud.ibm.com/docs/en/api/qiskit/circuit_random#random_clifford_circuit) ?

debasmita2102 commented:
Thanks for your feedback, @ShellyGarion . I am updating `BasicSimulator` so the max qubit limit for Clifford circuits is based on the tableau scaling you described (2N×2N elements). This should allow simulation of much larger Clifford circuits, potentially up to 2048 qubits with the current memory budget, while keeping the statevector limit at 24. I shall implement conditional logic and document the two pathway (Statevector simulation path and Clifford/Stabilizer simulation path) limits in the backend.

debasmita2102 commented:
> Note that the benchmark in this [comment](https://github.com/Qiskit/qiskit/pull/15159#issuecomment-3540772186), is a state with 2^{N-1} possible outputs. What happens if we take a GHZ circuit which has only 2 possible outputs? or big random clifford circuits (https://quantum.cloud.ibm.com/docs/en/api/qiskit/circuit_random#random_clifford_circuit) ?

Okay I will add new benchmarks as you suggested: one with a GHZ circuit (which only has two possible measurement outputs) and another with randomly generated Clifford circuits using the Qiskit API link you provided and compare both runtime and output distributions between statevector simulation and Clifford optimization, for small and large qubit counts and will share these benchmark results.

debasmita2102 committed:
Add BasicSimulator Clifford limits, tests, and GHZ/random Clifford benchmarks

debasmita2102 commented:
@ShellyGarion Based on the previous discussion:

**1. Dual qubit limits and branching**

BasicSimulator now uses separate limits:

`MAX_QUBITS_STATEVECTOR = 24` for the statevector path.

`MAX_QUBITS_CLIFFORD = 2048 `for the Clifford/Stabilizer path, based on the tableau scaling you described.​

**2. `_validat`e picks the limit per circuit:**

If `use_clifford_optimization `is `True` and `_is_clifford_circuit(circuit) `returns a Clifford, it applies the Clifford limit.

Otherwise it applies the 24‑qubit statevector limit.

**3. New unit tests cover:**

Statevector errors for >24 qubits.

Non‑Clifford circuits with `use_clifford_optimization=True `still hitting the statevector limit.

Clifford circuits above the statevector range that succeed with optimization enabled.

Error messages that mention the relevant limit/path (24 or Clifford).​

debasmita2102 commented:
@alexanderivrii 
Regarding Benchmarks: GHZ and random Clifford -  I added two ASV benchmarks in test/benchmarks/basic_simulator_clifford_benchmarks.py:

BasicSimulatorGHZBenchmark: GHZ‑like Clifford circuit (only 2 possible outputs), measured for n_qubits = [4, 8, 12, 16].


                ========== =========
              n_qubits    Clifford Statevector     
             ---------- --------- ---------- 

                 4       387±0ms        1.17±0ms 
                 8       1.83±0s           1.87±0ms 
                 12      5.66±0s          3.32±0ms 
                 16      11.8±0s           15.2±0ms 
             ========== =========
             
       
BasicSimulatorRandomCliffordBenchmark: random Clifford circuits generated via random_clifford(n_qubits) and converted to QuantumCircuit, also for n_qubits = [4, 8, 12, 16]

               ========== =========
              n_qubits  Clifford Statevector            
             ---------- ---------  ---------
                 4       447±0ms  1.40±0ms 
                 8       2.11±0s       2.96±0ms 
                 12      6.35±0s    11.7±0ms 
                 16      16.3±0s     157±0ms 
             ========== =========
             
             
  In both benchmarks, the current StabilizerState‑based Clifford path is noticeably slower than the statevector path in the 4–16 qubit range, even for GHZ circuits with only two output bitstrings. This aligns with the earlier point that performance work on the stabilizer backend should likely be handled separately (e.g. in Rust) rather than within this PR, while still justifying the higher theoretical Clifford qubit limit in the interface.​

debasmita2102 committed:
Fix pylint issues in BasicSimulator benchmarks

debasmita2102 committed:
Add Qiskit copyright header to benchmarks

debasmita2102 committed:
Merge remote-tracking branch 'upstream/main' into fix/issue-13610-stabilizer-backend

alexanderivrii commented on test/benchmarks/basic_simulator_clifford_benchmarks.py:
```diff
@@ -0,0 +1,92 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2017, 2023.
```
Minor:

```suggestion
# (C) Copyright IBM 2025.
```

alexanderivrii commented on benchmark_performance.py:
```diff
@@ -0,0 +1,68 @@
+#!/usr/bin/env python3
+"""Benchmark BasicSimulator performance for Clifford and non-Clifford circuits."""
+
```
You are not planning to commit this file, right? As you already know, we typically track performance using ASV.

alexanderivrii commented on benchmark_results.txt:
```diff
@@ -0,0 +1,55 @@
+======================================================================
+BasicSimulator Performance Benchmark - Clifford vs Non-Clifford
+======================================================================
```
You are not planning to commit this file, right? As you already know, we typically track performance using ASV.

debasmita2102 replied:
My mistake..

alexanderivrii commented on qiskit/providers/basic_provider/basic_simulator.py:
```diff
@@ -66,9 +77,12 @@ class BasicSimulator(BackendV2):
     # which matches the ~268 MB of required memory.
-    MAX_QUBITS_MEMORY = 24
+
+    MAX_QUBITS_STATEVECTOR = 24  # For statevector
+    MAX_QUBITS_CLIFFORD = 2048  # For Clifford/Stabilizer simulation
```
These are only used in `_validate` right now? Do we run all the validation checks by default? If yes, are they sufficiently fast (or should we put them under an option)?

debasmita2102 replied:
Yes, both `MAX_QUBITS_STATEVECTOR` and `MAX_QUBITS_CLIFFORD` are only used inside` _validate`. `_validate` is called once per circuit and just does a couple of integer comparisons on `circuit.num_qubits` to decide which limit applies and whether to raise an error. There’s no Clifford construction or other heavy work happening there, so the cost is negligible compared to either the `statevector` or `stabilizer `simulation. Because these checks prevent running circuits that would exceed the available memory, keeping them on by default felt safer than putting them behind an option. But let me know if you think otherwise.

alexanderivrii replied:
Thanks! Technically `_validate` does call `_is_clifford_circuit`, which construct a Clifford from a circuit and might be potentially slow, however in the quick benchmarking I did earlier that was not the main bottleneck, and I would be happy to accept the PR as is. But let's allow Shelly to take another look.

alexanderivrii reviewed (commented):
Thanks for adding the ASV tests, it will definitely help us track potential progress to the Clifford class. I have left a few questions.

debasmita2102 committed:
Remove ad-hoc BasicSimulator benchmark scripts

debasmita2102 committed:
Update copyright header year in Clifford benchmarks

debasmita2102 committed:
Merge branch 'main' into fix/issue-13610-stabilizer-backend

debasmita2102 committed:
Merge branch 'main' into fix/issue-13610-stabilizer-backend

alexanderivrii commented:
@debasmita2102, it seems that two of the new python tests that you have added, namely `test_unitary` and `test_statevector_qubit_limit_pass`  are very slow, leading to a time-out.

debasmita2102 commented:
> @debasmita2102, it seems that two of the new python tests that you have added, namely `test_unitary` and `test_statevector_qubit_limit_pass` are very slow, leading to a time-out.

@alexanderivrii   I’ve made lighter versions of test_unitary and test_statevector_qubit_limit_pass by reducing the qubit counts and number of shots, but keeping the same logic and assertions, so they still test the unitary application and the statevector success path while running much faster and avoiding CI timeouts. Is this approach acceptable, or would you prefer a different way to slim these tests down?

ShellyGarion commented:
> > @debasmita2102, it seems that two of the new python tests that you have added, namely `test_unitary` and `test_statevector_qubit_limit_pass` are very slow, leading to a time-out.
> 
> @alexanderivrii I’ve made lighter versions of test_unitary and test_statevector_qubit_limit_pass by reducing the qubit counts and number of shots, but keeping the same logic and assertions, so they still test the unitary application and the statevector success path while running much faster and avoiding CI timeouts. Is this approach acceptable, or would you prefer a different way to slim these tests down?

I think that there may be a problem with the code and not with the tests. The tests should represent a normal behaviour of the users, and it seems that now in some of the cases the `BasicSimulator` behaviour becomes exteremly slow. 
`test_unitary` is an old tests that constructs a simple 4-qubit circuit with 1024 shots, it shouldn't be that slow.
I suspect that the problem with `test_unitary` is that it adds X-gates as matrices, and it take time to verify that these are Cliffords and convert them.
`test_statevector_qubit_limit_pass` is a new one and outputs a 2^24 elements vector. This can indeed be too much to handle and the number of qubits can be reduced here, or replacing the test by another GHZ-like on 24 qubits but with a smaller output.

debasmita2102 committed:
Restore fast default BasicSimulator path and lighten statevector limit test

debasmita2102 committed:
Format BasicSimulator tests with black

alexanderivrii commented:
Sorry for the delayed reply. I agree with @ShellyGarion that we should not change the existing tests (and if we do see significant slowdowns, we should try to address them). However, I have now rerun the original `test_unitary` test with your changes, and it seems quite fast, could you also double-check and restore it?

The other minor thing (which, haha, Matthew always points out to me) is that's better to separate adding ASV benchmarks into a separate PR.

alexanderivrii commented on test/python/providers/basic_provider/test_basic_simulator.py:
```diff
@@ -288,6 +292,217 @@ def test_options(self):
                 else:
                     np.testing.assert_array_equal(getattr(out_options, key), in_options.get(key))
 
+    def test_clifford_circuit_bell_state(self):
+        """Test that Clifford circuits use StabilizerState backend."""
+        # Bell state is a Clifford circuit
```
One additional question: how can we see that StabilizerState backend is used?

debasmita2102 replied:
I updated `test_clifford_circuit_bell_state` to always run with` use_clifford_optimization=True`, so this Bell circuit is explicitly executed through the stabilizer path when it is detected as Clifford.

To avoid flakiness from random sampling, the test now only asserts that the outcomes are restricted to the Bell bitstrings (00 and 11) and that no 01 or 10 appear, while still checking that the total number of shots is correct. This keeps the test robust even if, for a particular seed, all shots land in just one of the Bell outcomes.

alexanderivrii commented on test/python/providers/basic_provider/test_basic_simulator.py:
```diff
@@ -288,6 +292,217 @@ def test_options(self):
+    def test_large_clifford_circuit_performance(self):
+        """Test that large Clifford circuits can be simulated efficiently."""
+        # Create a larger Clifford circuit (would be slow with statevector)
+        num_qubits = 10
```
Hmm, with the decreased value of `num_qubits`, the comment above does not seem very accurate. Similar to my previous question: can we somehow know which backend actually runs?

debasmita2102 replied:
I have updated the tests so that we use circuits that are above the 24‑qubit statevector limit but below the Clifford limit, and run them with `use_clifford_optimization=True`. In these cases the job succeeds only when the Clifford/StabilizerState path is selected, so the fact that the 32‑qubit Clifford tests pass implicitly shows that the stabilizer backend is actually being used.

alexanderivrii replied:
I still see `num_qubits=10`. I agree that changing this number to `32` is a good idea as this indeed checks that the StabilizerState path is selected. 

Update: actually there is also a `test_clifford_qubit_limit_pass` that checks that StabilizerState path is selected, so I a fine either way here.

alexanderivrii commented on test/python/providers/basic_provider/test_basic_simulator.py:
```diff
@@ -288,6 +292,217 @@ def test_options(self):
+        sim = BasicSimulator()
+        qc = QuantumCircuit(25)
+        with self.assertRaises(BasicProviderError):  # Or BasicProviderError
+            sim.run(qc, use_clifford_optimization=False)
```
Do we need the comment ``# Or BasicProviderError``?

debasmita2102 replied:
deleted.

alexanderivrii commented on test/python/providers/basic_provider/test_basic_simulator.py:
```diff
@@ -288,6 +292,217 @@ def test_options(self):
+        """Should succeed for a small Clifford circuit when optimization is on."""
+        sim = BasicSimulator()
+        n_qubits = 10
+        qc = QuantumCircuit(n_qubits)
```
Can we change `n_qubits` to `32` here? It would be nice to have some tests to show value of StabilizerState-based approach?

debasmita2102 replied:
updated n_qubits to 32

debasmita2102 committed:
Merge remote-tracking branch 'upstream/main' into fix/issue-13610-stabilizer-backend
Merge upstream/main into fix/issue-13610-stabilizer-backend

debasmita2102 committed:
Restore original test_unitary from main

debasmita2102 committed:
Relax Bell-state Clifford test and keep BasicSimulator tests passing

debasmita2102 commented:
> However, I have now rerun the original `test_unitary` test with your changes, and it seems quite fast, could you also double-check and restore it?


@alexanderivrii  I’ve restored test_unitary to the original 4‑qubit, 1024‑shot version from main, as you suggested.

debasmita2102 committed:
Remove leftover BasicProviderError comment and bump n_qubits to 32

debasmita2102 commented:
Thanks @alexanderivrii @ShellyGarion for the detailed feedback. I have cleaned up the remaining comments and qubit settings. All `test_basic_simulator`.py tests are passing locally. please let me know if you would like any further tweaks

alexanderivrii commented on test/python/providers/basic_provider/test_basic_simulator.py:
```diff
@@ -191,7 +191,7 @@ def test_memory(self):
     def test_unitary(self):
         """Test unitary gate instruction"""
-        max_qubits = 4
+        max_qubits = 4  # changed back to 4
```
Thanks for doing this. No need for explicit comment though.

```suggestion
        max_qubits = 4
```

alexanderivrii commented on test/python/providers/basic_provider/test_basic_simulator.py:
```diff
@@ -201,7 +201,7 @@ def test_unitary(self):
                 multi_x = np.kron(multi_x, x_mat)
             # Target counts
-            shots = 1024
+            shots = 1024  # changed back to 10124
```
Same here.

```suggestion
            shots = 1024
```

alexanderivrii reviewed (commented):
Thanks for the changes, @debasmita2102. Two super-minor nitpicks about the comments + one optional suggestion to increase the number of qubits in one of the tests from `10` to `32` as per your comment. After that it's ready to go.

Oh, I was also asking in the previous review if you would like to move new ASV benchmarks to a separate PR. I am fine either way, but we usually do try to separate the two things.

debasmita2102 commented:
> Thanks for the changes, @debasmita2102. Two super-minor nitpicks about the comments + one optional suggestion to increase the number of qubits in one of the tests from `10` to `32` as per your comment. After that it's ready to go.
> 
> Oh, I was also asking in the previous review if you would like to move new ASV benchmarks to a separate PR. I am fine either way, but we usually do try to separate the two things.

I will resolve the typos and missing edits.. and will shift the asv too! thanks a lot

debasmita2102 committed:
Update test/python/providers/basic_provider/test_basic_simulator.py

Co-authored-by: Alexander Ivrii <alexi@il.ibm.com>

debasmita2102 committed:
Update test/python/providers/basic_provider/test_basic_simulator.py

Co-authored-by: Alexander Ivrii <alexi@il.ibm.com>

debasmita2102 committed:
Remove BasicSimulator ASV benchmarks from Clifford optimization PR and Ensure 32-qubit Clifford test uses stabilizer path

debasmita2102 commented:
The seperated ASV Benchmark PR #15448

alexanderivrii reviewed (approved):
Thanks for all the changes, LGTM!

alexanderivrii commented:
I have approved the PR, @ShellyGarion do you have any additional comments? Also, I suggest to merge #15448 first, see [this comment](https://github.com/Qiskit/qiskit/pull/15448#issuecomment-3660345486).

ShellyGarion commented:
> I have approved the PR, @ShellyGarion do you have any additional comments? Also, I suggest to merge #15448 first, see [this comment](https://github.com/Qiskit/qiskit/pull/15448#issuecomment-3660345486).

LGTM! Thanks @debasmita2102 for this contribution to Qiskit.
I tried running this code for a GHZ circuit with 2000 qubits, and it ran well (and fast!)
I also agree that it's ready to being merged now.

Cryoris commented on qiskit/providers/basic_provider/basic_simulator.py:
```diff
@@ -27,6 +27,13 @@
 field, which is a result of measurements for each shot.
+
+
+# The simulator supports:
```
How does this render in the docs?

ShellyGarion replied:
I think this does not appear in the API docs

debasmita2102 replied:
This comment is in the module docstring of `basic_simulator`.py, but it is not part of the public class API and does not currently show up in the rendered API docs (as Shelly noted). If you prefer, I can either move this text into the class docstring or drop it from the docstring entirely and keep it as a regular code comment instead.

ShellyGarion replied:
I think you can replace it by a comment that looks like this:
` """ # The simulator supports: ... """`

Cryoris replied:
Instead of having comments here, could we just clearly document the behavior in the class docstring of `BasicSimulator` (rather than the module)?

debasmita2102 replied:
Thanks! Done in #15467

Cryoris commented on qiskit/providers/basic_provider/basic_simulator.py:
```diff
@@ -214,6 +231,7 @@ def _default_options(cls) -> Options:
             memory=True,
             initial_statevector=None,
             seed_simulator=None,
+            use_clifford_optimization=False,  # ADDED FOR CLIFFORD
```
We don't typically add comments like this (imagine we would add them for every PR, then the code would become crowded very quickly 🙂). If we do a potential follow up for the docs above, it would be nice to also remove this inline comments

ShellyGarion replied:
This is an internal comment and does not appear in the API docs (but if you want it can be removed in a follow-up)
<img width="1550" height="1208" alt="image" src="https://github.com/user-attachments/assets/161673ba-c383-4832-9a8b-477856f87d33" />

debasmita2102 replied:
I understand. I will check in future for this kind of messy comments.  This is already merged so no idea what to do now.

ShellyGarion replied:
you can always open another PR to fix it.

ShellyGarion commented on releasenotes/notes/clifford-optimization-basicsimulator-95099da6faf2ede3.yaml:
```diff
@@ -0,0 +1,30 @@
+---
+
+features:
```
should it be `features_providers` ?
since this PR is already merged, should we handle it now or as part of Qiskit 2.4 release?

jakelishman replied:
It should, yeah - if you'd like to open a PR to fix it now that'd make sure we don't miss it in the 2.4 clean-up.
